### DISCUSSION TITLE
PATH for exec commands $HOME/.local/bin and $HOME/bin are not used

### DISCUSSION THREAD

**Discussion by waterkip** [i3/i3#5928] [Q&A] [answered]
Hi all,

I have a question regarding the PATH that is used by i3. I am trying to "override" dmenu from suckless by rofi. Rofi allows that if you call rofi as dmenu, eg, `ln -s /usr/bin/rofi /usr/bin/dmenu`, that it should behave correctly. This works.

However, when I add `$HOME/bin/dmenu` ($HOME/bin is in my PATH) this doesn't get picked up.

To reproduce this, you can simply install a small Debian 12 VM with i3. Install suckless-tools (`apt-get install suckless-tools`). Create a file called `$HOME/bin/dmenu_run` with the contents:

```
echo I was called at $(date) >> dmenu.ran
```

For those who are wondering, why are you working with dmenu_run and not dmenu, well, `dmenu_run` calls `dmenu_path | dmenu .. `. The i3 exec call is to dmenu_run, if the PATH is respected, the dmenu call should be correct, which now isn't.

Exec the dmenu shortcut (mod+d) and notice that the dmenu.ran doesn't have any content. You will see that the dmenu from suckless is executed. Now in a terminal, run dmenu_run and see that dmenu.ran has contents in it so we know is has been executed.

I'm trying to figure out why this is. The [manual](https://i3wm.org/docs/userguide.html#exec) states:
> 6.1. Executing applications (exec)
> What good is a window manager if you can’t actually start any applications? The exec command starts
> an application by passing the command you specify to a shell. This implies that you can use
> globbing (wildcards), and programs will be searched in your $PATH.

FWIW, I'm running zsh, but the parts that set the relevant bits of my PATH are found in my `.profile`

What I find also interesting is that when I remove suckless-tools and the dmenu* files from /usr/bin are removed i3, in its log, don't show that the command cannot be found:

```
02/24/2024 11:35:31 AM - ../src/commands_parser.c:parse_command:241 - COMMAND: *exec --no-startup-id dmenu_run*
02/24/2024 11:35:31 AM - ../src/commands.c:cmd_criteria_init:148 - Initializing criteria, current_match = 0x5591f5fc53c0
02/24/2024 11:35:31 AM - ../src/commands.c:cmd_exec:1298 - should execute dmenu_run, no_startup_id = 1
02/24/2024 11:35:31 AM - executing: dmenu_run
02/24/2024 11:35:31 AM - ../src/commands.c:cmd_criteria_init:148 - Initializing criteria, current_match = 0x5591f5fc53c0
02/24/2024 11:35:31 AM - ../src/ipc.c:ipc_send_binding_event:1682 - Issue IPC binding run event (sym = d, code = 0)
02/24/2024 11:35:32 AM - ../src/handlers.c:handle_event:1384 - event type 3, xkb_base 85
02/24/2024 11:35:32 AM - ../src/key_press.c:handle_key_press:23 - KeyRelease 40, state raw = 0x40
02/24/2024 11:35:32 AM - ../src/bindings.c:get_binding_from_xcb_event:318 - (removed capslock, state = 0x40)
02/24/2024 11:35:32 AM - ../src/bindings.c:get_binding_from_xcb_event:343 - (transformed keyboard group, state = 0x10040)
```

When it is executed:
```
02/24/2024 11:39:28 AM - ../src/commands_parser.c:parse_command:241 - COMMAND: *exec --no-startup-id dmenu_run*
02/24/2024 11:39:28 AM - ../src/commands.c:cmd_criteria_init:148 - Initializing criteria, current_match = 0x5591f5fc53c0
02/24/2024 11:39:28 AM - ../src/commands.c:cmd_exec:1298 - should execute dmenu_run, no_startup_id = 1
02/24/2024 11:39:28 AM - executing: dmenu_run
02/24/2024 11:39:28 AM - ../src/commands.c:cmd_criteria_init:148 - Initializing criteria, current_match = 0x5591f5fc53c0
02/24/2024 11:39:28 AM - ../src/ipc.c:ipc_send_binding_event:1682 - Issue IPC binding run event (sym = d, code = 0)
02/24/2024 11:39:28 AM - ../src/handlers.c:handle_event:1384 - event type 3, xkb_base 85
02/24/2024 11:39:28 AM - ../src/key_press.c:handle_key_press:23 - KeyRelease 40, state raw = 0x40
02/24/2024 11:39:28 AM - ../src/bindings.c:get_binding_from_xcb_event:318 - (removed capslock, state = 0x40)
02/24/2024 11:39:28 AM - ../src/bindings.c:get_binding_from_xcb_event:343 - (transformed keyboard group, state = 0x10040)
```
It is both present in 4.22 (Debain stable) and 4.23 (Debian unstable)

```
# Debian stable:
Binary i3 version:  4.22 (2023-01-02) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.22 (2023-01-02) (pid 801)
Loaded i3 config:
  /home/wesleys/.config/i3/config (main) (last modified: Sat 23 Dec 2023 03:48:03 PM AST, 5428680 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3-with-/i3
i3 version 4.22 (2023-01-02) © 2009 Michael Stapelberg and contributors

# Debian unstable
Binary i3 version:  4.23 (2023-10-29) © 2009 Michael Stapelberg and contributors
Running i3 version: 4.23 (2023-10-29) (pid 124571)
Loaded i3 config:
  /home/wesleys/.config/i3/config (main) (last modified: Sat Feb 24 00:25:19 2024, 40881 seconds ago)

The i3 binary you just called: /usr/bin/i3
The i3 binary you are running: i3
i3 version 4.23 (2023-10-29) © 2009 Michael Stapelberg and contributors
```

**Comment by orestisfl** [comment]
> However, when I add $HOME/bin/dmenu ($HOME/bin is in my PATH) this doesn't get picked up.

You need to make sure it is the PATH i3 sees as well. You can follow https://unix.stackexchange.com/a/70636 for example

> FWIW, I'm running zsh, but the parts that set the relevant bits of my PATH are found in my .profile

Can you share that file? Are you sure it is sourced by your login manager / xinit?

> What I find also interesting is that when I remove suckless-tools and the dmenu* files from /usr/bin are removed i3, in its log, don't show that the command cannot be found

Yes, i3 executes commands via sh and doesn't check the output.

**Reply by waterkip** [reply]
> Can you share that file? Are you sure it is sourced by your login manager / xinit?

I can't say for sure.. Do you know how I can be sure they are sourced? Is there some way I can ask i3 this? nvm, `cat /proc/<pidofi3>/environ` tells me this:

```
PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
```

I think I need to dig in to lightdm perhaps?

On my testbed the .profile is:
```
vbox-debian% cat .profile
# ~/.profile: executed by the command interpreter for login shells.
# This file is not read by bash(1), if ~/.bash_profile or ~/.bash_login
# exists.
# see /usr/share/doc/bash/examples/startup-files for examples.
# the files are located in the bash-doc package.

# the default umask is set in /etc/profile; for setting the umask
# for ssh logins, install and configure the libpam-umask package.
#umask 022

# if running bash
if [ -n "$BASH_VERSION" ]; then
    # include .bashrc if it exists
    if [ -f "$HOME/.bashrc" ]; then
	. "$HOME/.bashrc"
    fi
fi

# set PATH so it includes user's private bin if it exists
if [ -d "$HOME/bin" ] ; then
    PATH="$HOME/bin:$PATH"
fi

# set PATH so it includes user's private bin if it exists
if [ -d "$HOME/.local/bin" ] ; then
    PATH="$HOME/.local/bin:$PATH"
fi
```

> Yes, i3 executes commands via sh and doesn't check the output.

k

**Reply by waterkip** [reply]
I have to go now, but I think this might be something I need to look into: 
https://bbs.archlinux.org/viewtopic.php?id=276407

**Reply by orestisfl** [reply]
I don't see an `export` there. For lightdm this has worked in other threads: https://github.com/i3/i3/discussions/5892#discussioncomment-8309608

**Reply by waterkip** [reply]
I added it, with a twist: `export PATH=$HOME/weird/bin:$PATH` but it's not picked up:

```
 PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
```

**Reply by orestisfl** [reply]
Just to be sure, you need to log-out and log-in for this to be picked up. Otherwise, you can try `~/.xprofile`.

**Reply by waterkip** [reply]
I think I found the answer, according to [Debian](https://wiki.debian.org/LightDM#Debian-specific_information):
> The Debian package, unlike the Ubuntu one, does not contain the /usr/bin/lightdm-session script loading [EnvironmentVariables](https://wiki.debian.org/EnvironmentVariables) from ~/.profile and other similar files. To get behavior similar to [GDM](https://wiki.debian.org/GDM) and [SDDM](https://wiki.debian.org/SDDM) override the session-wrapper setting with a script sourcing profile scripts before running /etc/X11/Xsession. For example to use a script from the [sddm](https://packages.debian.org/sddm) package create a file like /etc/lightdm/lightdm.conf.d/80-session-wrapper-sddm.conf

**Reply by waterkip** [reply, marked as answer]
So for anyone reading this who has Debian. You'll need to tweak the session-wrapper of LightDM. In `/usr/share/lightdm/lightdm.conf.d/01_debian.conf` the session-wrapper is changed from the default `/usr/bin/lightdm-session` to `/etc/X11/Xsession` (which is provided by `x11-common`, via `suckless-tools`). You'll want to override this, you can do this either by adding a configuration item in `/usr/share/lightdm/lightdm.conf.d/02_yourtweaks.conf` or in `/etc/lightdm/lightdm.conf`. Be advised that you can either set `[SeatDefault]` or `[Seat:*]`, the former is a deprecated variant btw.

Now you can create a separate wrapper script, for my testing purposes I've added a very minimal script:
```
#!/usr/bin/sh

for file in "/etc/profile" "$HOME/.profile" "/etc/xprofile" "$HOME/.xprofile"; do
    [ ! -f "$file" ] && continue
    . $file
done
```

You can look at [Ubuntu's version](https://github.com/canonical/lightdm/blob/main/debian/lightdm-session) or a similar script that is provided by [sddm from Debian](https://salsa.debian.org/qt-kde-team/3rdparty/sddm/-/blob/master/debian/Xsession?ref_type=heads) for more advanced versions of a wrapper script.

Now your `.profile`/`.xprofile` is loaded and your PATH is setup correctly.
